test/functional/pdo_sqlsrv/pdostatement_format_decimals.phpt

```diff
@@ -0,0 +1,385 @@
+5. Do not support output params
+--ENV--
+PHPT_EXEC=true
+--SKIPIF--
+<?php require('skipif_mid-refactor.inc'); ?>
+--FILE--
+<?php
+require_once("MsCommon_mid-refactor.inc");
+
+function checkException($exception, $expected)
+{
+    if (strpos($exception->getMessage(), $expected) === false) {
+        print_r($exception->getMessage());
+        echo "\n";
+    }
+}
+
+function testPdoAttribute($conn, $setAttr)
+{
+    // Expects exception because PDO::SQLSRV_ATTR_FORMAT_DECIMALS
+    // is a statement level attribute
+    try {
+        $res = true;
+        if ($setAttr) {
+            $res = $conn->setAttribute(PDO::SQLSRV_ATTR_FORMAT_DECIMALS, 1);
+        } else {
+            $res = $conn->getAttribute(PDO::SQLSRV_ATTR_FORMAT_DECIMALS);
+        }
+        if ($res) {
+            echo "setAttribute at PDO level should have failed!\n";
+        }
+    } catch (PdoException $e) {
+        if ($setAttr) {
+            $expected = 'The given attribute is only supported on the PDOStatement object.';
+        } else {
+            $expected = 'driver does not support that attribute';
+        }
+
+        checkException($e, $expected);
+    }
+}
+
+function testErrorCases($conn)
+{
+    $query = "SELECT 0.0001";
+
+    try {
+        $options = array(PDO::SQLSRV_ATTR_FORMAT_DECIMALS => 0.9);
+        $stmt = $conn->prepare($query, $options);
+    } catch (PdoException $e) {
+        $expected = 'Expected an integer to specify number of decimals to format the output values of decimal data types';
+        checkException($e, $expected);
+    }
+
+    try {
+        $options = array(PDO::SQLSRV_ATTR_FORMAT_DECIMALS => 100);
+        $stmt = $conn->prepare($query, $options);
+    } catch (PdoException $e) {
+        $expected = 'For formatting decimal data values, 100 is out of range. Expected an integer from 0 to 38, inclusive.';
+        checkException($e, $expected);
+    }
+}
+
+function verifyMoneyValues($conn, $query, $values, $numDigits)
+{
+    $options = array(PDO::SQLSRV_ATTR_FORMAT_DECIMALS => $numDigits);
+    $stmt = $conn->prepare($query, $options);
+    $stmt->execute();
+    $results = $stmt->fetch(PDO::FETCH_NUM);
+
+    trace("\nverifyMoneyValues:\n");
+    for ($i = 0; $i < count($values); $i++) {
+        $value = number_format($values[$i], $numDigits);
+        trace("$results[$i], $value\n");
+
+        if ($value !== $results[$i]) {
+            echo "testMoneyTypes: Expected $value but got $results[$i]\n";
+        }
+    }
+}
+
+function testFloatTypes($conn)
+{
+    // This test with the float types of various number of bits, which are retrieved
+    // as numbers by default. When fetched as strings, no formatting is done even with
+    // the statement option FormatDecimals set
+    $epsilon = 0.001;
+    $values = array();
+    for ($i = 0; $i < 5; $i++) {
+        $n1 = rand(1, 100);
+        $n2 = rand(1, 100);
+        $neg = ($i % 2 == 0) ? -1 : 1;
+
+        $n = $neg * $n1 / $n2;

```

Comment:
This is going to do integer multiplication and division - it's not clear if that's what we want for this test, won't this just be zero half the time (because of the rand())?

Reply:
No, actually the results are floats, not integers. For example, running the test in debug mode gave me these values

```
testFloatTypes: -0.059999999, -0.059999999
testFloatTypes: 1.7391304, 1.7391304
testFloatTypes: -3.2333333, -3.2333333
testFloatTypes: 0.048780487804878, 0.048780487804878
testFloatTypes: -0.56701030927835, -0.56701030927835
```